src/core/execute.c

```diff
@@ -2161,10 +2161,18 @@ static int setup_exec_directory(
                         if (r < 0 && r != -EEXIST)
                                 goto fail;
                         if (r == -EEXIST) {
-                                if (chmod(p, context->directories[type].mode) < 0) {
+                                struct stat st;
+                                mode_t mask = 07777;
+
+                                if (stat(p, &st) < 0) {
                                         r = -errno;
                                         goto fail;
                                 }
+                                if (((st.st_mode ^ context->directories[type].mode) & mask ) != 0) {
```

Comment:
Please drop a whitespace between `mask` and `)`.
Also, please drop unnecessary `{}`.